lib/data-types.js

```diff
@@ -1,281 +1,489 @@
-  if (this instanceof STRING) {
-    this._binary = !!binary;
-    if (typeof length === 'number') {
-      this._length = length;
-    } else {
-      this._length = 255;
-    }
-  } else {
-    return new STRING(length, binary);
-  }
+var util = require('util');
+
+var ABSTRACT = function(options) {
+
 };
 
-var CHAR = function(length, binary) {
-  if (this instanceof CHAR) {
-    this._binary = !!binary;
-    if (typeof length === 'number') {
-      this._length = length;
-    } else {
-      this._length = 255;
-    }
-  } else {
-    return new CHAR(length, binary);
-  }
+ABSTRACT.prototype.toString = function() {
+  return this.toSql();
+};
+ABSTRACT.prototype.toSql = function() {
+  return this.key;
+};
+
+/**
+ * A variable length string. Default length 255
+ *
+ * Available properties: `BINARY`
+ *
+ * @property STRING
+ */
+var STRING = function(length, binary) {
+  var options = typeof length === "object" && length || {
+    length: length,
+    binary: binary
+  };
+
+  if (!(this instanceof STRING)) return new STRING(options);
+
+  this.options = options;
+  this._binary = options.binary;
+  this._length = options.length || 255;
 };
+util.inherits(STRING, ABSTRACT);
 
-STRING.prototype = {
-  get BINARY() {
+STRING.prototype.key = STRING.key = 'STRING';
+STRING.prototype.toSql = function() {
+  return 'VARCHAR(' + this._length + ')' + ((this._binary) ? ' BINARY' : '');
+};
+Object.defineProperty(STRING.prototype, 'BINARY', {
+  get: function() {
     this._binary = true;
     return this;
-  },
-  get type() {
-    return this.toString();
-  },
-  toString: function() {
-    return 'VARCHAR(' + this._length + ')' + ((this._binary) ? ' BINARY' : '');
   }
+});
+
+/**
+ * A fixed length string. Default length 255
+ *
+ * Available properties: `BINARY`
+ *
+ * @property CHAR
+ */
+var CHAR = function(length, binary) {
+  var options = typeof length === "object" && length || {
+    length: length,
+    binary: binary
+  };
+
+  if (!(this instanceof CHAR)) return new CHAR(options);
+  STRING.apply(this, arguments);
 };
+util.inherits(CHAR, STRING);
 
-CHAR.prototype = {
-  get BINARY() {
-    this._binary = true;
-    return this;
-  },
-  get type() {
-    return this.toString();
-  },
-  toString: function() {
-    return 'CHAR(' + this._length + ')' + ((this._binary) ? ' BINARY' : '');
+CHAR.prototype.key = CHAR.key = 'CHAR';
+CHAR.prototype.toSql = function() {
+  return 'CHAR(' + this._length + ')' + ((this._binary) ? ' BINARY' : '');
+};
+
+
+/**
+ * An unlimited length text column
+ * @property TEXT
+ */
+var TEXT = function(options) {
+  if (!(this instanceof TEXT)) return new TEXT(options);
+};
+util.inherits(TEXT, ABSTRACT);
+
+TEXT.prototype.key = TEXT.key = 'TEXT';
+
+
+var NUMBER = function(options) {
+  if (!(this instanceof NUMBER)) return new NUMBER(options);
+  this.options = options;
+  this._length = options.length;
+  this._decimals = options.decimals;
+  this._precision = options.precision;
```

Comment:
@mbroadst one is for DECIMAL other is for FLOAT i think. Thats pure conversion from the old code